src/routes/did.rs

```diff
@@ -0,0 +1,221 @@
+        let now = chrono::Utc::now().timestamp();
+        if self.exp < now && self.nbf > now {
+            return Err(Error::VerifyJWT("Token expired".to_string()));
+        }
+        Ok(())
+    }
+
+    fn check_iss_is_sub(&self) -> Result<(), Error> {
+        if self.iss != self.sub {
+            return Err(Error::VerifyJWT(
+                "Identity mismatch: iss is not sub".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[post("/api/v1/did/{token}")]
+async fn login_with_did(
+    app_state: web::Data<RwLock<AppState>>,
+    session: Session,
+    path: web::Path<String>,
+) -> Result<HttpResponse, Error> {
+    log::info!("POST DID handler");
+
+    let jwt_token = path.into_inner();
+
+    // get requirements for this client, the client id comes from the session
+    let oidc_context = session
+        .get::<AuthorizeQueryParameters>(OIDC_SESSION_KEY)
+        .map_err(|_| Error::OauthNoSession)?
+        .ok_or(Error::OauthInvalidClientId)?;
+
+    let nonce = oidc_context.nonce;
+
+    let mut hasher = Sha512::new();
+
+    let parts: Vec<&str> = jwt_token.split('.').collect();
+    let header = parts
+        .first()
+        .ok_or(Error::VerifyJWT("JWT header is missing".to_string()))?;
+    let body = parts
+        .get(1)
+        .ok_or(Error::VerifyJWT("JWT body is missing".to_string()))?;
+    let signature = parts
+        .get(2)
+        .ok_or(Error::VerifyJWT("JWT signature is missing".to_string()))?;
+
+    hasher.update(format!("{}.{}", header, body));
+    let data_to_verify_hex = hex_encode(hasher.finalize());
+    let data_to_verify = data_to_verify_hex.trim_start_matches("0x").as_bytes();
+
+    let decoded_header = general_purpose::STANDARD
+        .decode(header)
+        .map_err(|_| Error::VerifyJWT("Failed to decode header".to_string()))?;
+    let decoded_body = general_purpose::STANDARD
+        .decode(body)
+        .map_err(|_| Error::VerifyJWT("Failed to decode body".to_string()))?;
+    let decoded_signature = general_purpose::STANDARD
+        .decode(signature)
+        .map_err(|_| Error::VerifyJWT("Failed to decode signature".to_string()))?;
+
+    let jwt_header: JWTHeader = serde_json::from_slice(&decoded_header)
+        .map_err(|_| Error::VerifyJWT("Failed to parse header".to_string()))?;
+    let jwt_payload: JWTPayload = serde_json::from_slice(&decoded_body)
+        .map_err(|_| Error::VerifyJWT("Failed to parse payload".to_string()))?;
+
+    jwt_payload.check_is_expired()?;
+    jwt_payload.check_iss_is_sub()?;
+
+    let sender = &jwt_payload.iss;
+
+    let endpoint = {
+        let app_state = app_state.read()?;
+        app_state.kilt_endpoint.clone()
+    };
+    let cli = kilt::connect(&endpoint)
+        .await
+        .map_err(|_| Error::CantConnectToBlockchain)?;
+
+    let did_document = get_did_doc(sender, &cli).await?;
+
+    let signed_key = hex_decode(jwt_header.kid.trim_start_matches("0x"))
```

Comment:
I didn't catch that in the first review: The kid needs to be the full kid of the key used to sign the payload. Therefore it must contain the DID that signed it. 
It has to look like `kid: "did:kilt:4abcdefg#0xdeadbeaf112233"`. 
In the current code you just get the DID doc of the DID which is listed as sub and iss, and check that the key in kid is in that doc, so this is fine, but still not really compliant

Reply:
updated in: [bb1fb58](https://github.com/KILTprotocol/opendid/pull/28/commits/bb1fb58e552c2bfba19d7c7e8f0e27d0e0e4957b)